File: DSP/CrossFade.cpp
#include "CrossFade.hpp"
#include <cmath>
#include <algorithm>

namespace dsp {

void crossfade(const float* a, const float* b, float progress, float* out, size_t count) {
    if (!a || !b || !out) return;
    progress = std::clamp(progress, 0.0f, 1.0f);
    float inv = 1.0f - progress;
    for (size_t i = 0; i < count; ++i) out[i] = inv * a[i] + progress * b[i];
}

void crossfadeEqualPower(const float* a, const float* b, float progress, float* out, size_t count) {
    if (!a || !b || !out) return;
    progress = std::clamp(progress, 0.0f, 1.0f);
    float gA = std::cos(progress * 1.57079632679f); // cos(p * pi/2)
    float gB = std::sin(progress * 1.57079632679f);
    for (size_t i = 0; i < count; ++i) out[i] = gA * a[i] + gB * b[i];
}

void CrossfadeState::start(int numFrames) {
    totalFrames = numFrames;
    currentFrame = 0;
    active = (numFrames > 0);
}

float CrossfadeState::progress() const {
    if (!active || totalFrames == 0) return 1.0f;
    return std::clamp(static_cast<float>(currentFrame) / totalFrames, 0.0f, 1.0f);
}

void CrossfadeState::advance() {
    if (!active) return;
    ++currentFrame;
    if (currentFrame >= totalFrames) active = false;
}

void CrossfadeState::reset() {
    totalFrames = 0;
    currentFrame = 0;
    active = false;
}

} // namespace dsp

File: Tests/dsp/crossfade.py
import numpy as np
import math

def crossfade(a: np.ndarray, b: np.ndarray, progress: float) -> np.ndarray:
    p = float(np.clip(progress, 0, 1))
    return (1 - p) * a + p * b

def crossfade_equal_power(a: np.ndarray, b: np.ndarray, progress: float) -> np.ndarray:
    p = float(np.clip(progress, 0, 1))
    g_a = math.cos(p * math.pi / 2)
    g_b = math.sin(p * math.pi / 2)
    return g_a * a + g_b * b

class CrossfadeState:
    def __init__(self):
        self.total_frames = 0
        self.current = 0
        self.active = False
    def start(self, num_frames: int):
        self.total_frames = num_frames
        self.current = 0
        self.active = num_frames > 0
    def progress(self) -> float:
        if not self.active or self.total_frames == 0:
            return 1.0
        return max(0.0, min(1.0, self.current / self.total_frames))
    def advance(self):
        if not self.active:
            return
        self.current += 1
        if self.current >= self.total_frames:
            self.active = False
    def reset(self):
        self.total_frames = 0; self.current = 0; self.active = False
